### PR TITLE
Implemented page-break-before/after for Stylo

### PR DESCRIPTION
**Pull request by shinglyu**
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12442)

### CODE REVIEW

File: ports/geckolib/properties.mako.rs [hunk 1]
```diff
@@ -844,6 +844,40 @@ fn static_assert() {
...
+            T::always => true,
+            T::avoid  => false,
+            T::left   => true,
+            T::right  => true
+        };
+        // TODO(shinglyu): Rename Gecko's struct to mPageBreakBefore
+        self.gecko.mBreakBefore = result;
+    }
+
+    fn copy_page_break_before_from(&mut self, other: &Self) {
```
Comment (emilio) [inline ports/geckolib/properties.mako.rs]: nit: This could use the `impl_simple_copy` helper.

File: ports/geckolib/properties.mako.rs [hunk 2]
```diff
@@ -844,6 +844,40 @@ fn static_assert() {
     fn copy__moz_binding_from(&mut self, other: &Self) {
         unsafe { Gecko_CopyMozBindingFrom(&mut self.gecko, &other.gecko); }
     }
+
+    fn set_page_break_before(&mut self, v: longhands::page_break_before::computed_value::T) {
+        use style::computed_values::page_break_before::T;
+        let result = match v {
```
Comment (emilio) [inline ports/geckolib/properties.mako.rs]: hm... how does gecko discern between different values of the property if this is just a boolean for them? There should be something else to check it seems, could you elaborate?

Comment (emilio) [inline ports/geckolib/properties.mako.rs]: http://searchfox.org/mozilla-central/source/layout/style/nsRuleNode.cpp#5970

**Comment by emilio** [comment]
Could you point to the same bugzilla bug and section of the CSS spec as the comment in the Gecko file?

r=me with that and with or without the helper (preferably with).

Thanks! :)

File: ports/geckolib/properties.mako.rs (hunk 2 above)
Comment (shinglyu) [inline ports/geckolib/properties.mako.rs]: You got it. I'll add a comment here to clarify.

**Comment by emilio** [comment]
Travis failure doesn't seem related to this:

@bors-servo: r+

**Comment by frivoal** [comment]
Hi there: page-break-before and page-break-after are historical names, and they're now supposed to be low-profile aliases for the real thing: break-before and break-after.
https://drafts.csswg.org/css-break/#break-between
https://drafts.csswg.org/css-break/#page-break-properties

It would be nice if servo / stylo could get this right from the get go.

**Comment by emilio** [comment]
@frivoal: Thanks for the heads-up!

Servo doesn't implement the page break properties yet, and stylo had those two, that would need to become an alias it seems. I don't know off-hand how good our aliasing support is right now, but we definitely will need to support those properties (and more aliases) soon, so when we have the infra should be easy to do.

Thanks again! :)

**Comment by frivoal** [comment]
In this case at least, the aliasing mechanism should be a non issue, because https://drafts.csswg.org/css-break/#page-break-properties defines it to work by making page-break-* a shorthand of break-*.

There was quite a bit of discussion in the CSSWG a few years back (with some incomplete traces of the background research to be found here https://wiki.csswg.org/ideas/aliasing#shorthandlonghand-take-2), leading to the conclusion that making aliased properties shorthands of the canonical name was in general the best way to go, both because it lead to good behavior for cascading and the OM, and because it relied on a mechanism everybody already had, giving us a good chance that it would be done correctly and consistently across vendors.

**Pull request merged**